File: cpp/204.(哈希)Count Primes.cpp
//Count the number of prime numbers less than a non-negative number, n.

class Solution {
public:
    int countPrimes(int n) {
        if(n<2)
            return 0;
        int limit=sqrt(n), count=0;
        vector<bool> prime(n, true);
        for(int i=2; i<n; i++)
        {
            if(prime[i])
                ++count;
            if(i>limit)
                continue;
            for(int j=i*i; j<n; j+=i)
                prime[j]=false;
        }
        return count;
    } 
};

File: golang/204.(哈希)Count_Primes.go
// package golang
// name 204.(哈希)Count_Primes

package golang

import (
	"math"
)

func CountPrimes(n int) (count int) {
	if n < 2 {
		return 0
	}

	limit := int(math.Sqrt(float64(n)))
	count = 0
	primes := make([]bool, n, n)

	for i := 0; i < len(primes); i++ {
		primes[i] = true
	}

	for i := 2; i < n; i++ {
		if v := primes[i]; v == true {
			count++
		}
		if i > limit {
			continue
		}
		for j := i * i; j < n; j += i {
			primes[j] = false
		}
	}
	return count
}
